cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactionService.java

```diff
@@ -258,6 +288,8 @@ public Transaction getTransaction() {
               tx = transaction = Optional.of(txClient.startLong());
             } catch (Throwable t) {
               LOG.error("Failed to start transaction for job {}", jobId, t);
+              // Set the transaction to an absent Optional to indicate the starting of transaction failed.
+              // This will prevent future call to this method to attempt to start a transaction again
```

Comment:
There are three states for the `Optional<Transaction> transaction`:

  1. `transaction == null`. This means there is no transaction to use and there was no attempt to start a transaction. Hence and attempt will be made to start a transaction (line 256-258)
  2. `transaction != null` and the `Optional` is not absent. This mean there is a transaction associated with the job and we can use it for the job.
  3. `transaction != null` but the `Optional` is absent. This mean an attempt was made earlier to start a transaction, but failed. In this case, we won't try to start a transaction again. This is to make sure later stages of the same job fail to perform any transaction operation instead of performing it with a different transaction.